### PR TITLE
feat: add program owned Merkle tree checks

### PR DESCRIPTION
**Pull request by ananas-block**
Motivation:
- to ensure parallelism we need Merkle trees that can only be used by specific programs -> program owned
- the delegate field (to be renamed to program owner) is used to store the owning program pubkey

Changes:
- added checks state and adddress Merkle tree ownership to append_state, insert_into_nullifier_queue, and insert address functions in compressed pda
- refactored the signer checks into:
  - cpi_signer_check
  - input_compressed_accounts_signer_check
  - output_compressed_accounts_write_access_check

### CODE REVIEW

File: js/stateless.js/src/constants.ts [hunk 1]
```diff
@@ -74,6 +74,13 @@ export const TRANSACTION_MERKLE_TREE_ROLLOVER_THRESHOLD = new BN(
     Math.floor(2 ** DEFAULT_MERKLE_TREE_HEIGHT * 0.95),
 );
 
+/**
+ * Tip to reward the forrester that empties the queues and nullifies leaves.
```
Comment (SwenSchaeferjohann) [inline js/stateless.js/src/constants.ts:78]:
```suggestion
 * Tip to reward the forester that empties the queues and nullifies leaves.
```

File: programs/compressed-pda/src/instructions.rs [hunk 2]
```diff
@@ -31,18 +32,28 @@ pub fn process_execute_compressed_transaction<'a, 'b, 'c: 'info, 'info>(
...
     )?;
+    output_compressed_accounts_write_access_check(&inputs, &ctx.accounts.invoking_program)?;
 
     if let Some(cpi_context) = cpi_context {
         if process_cpi_context(cpi_context, &mut ctx, &mut inputs)?.is_some() {
             return Ok(());
         }
     }
+    // if invoking program is present then cpi context must be present
+    // TODO: move to plausiblility check
```
Comment (SwenSchaeferjohann) [inline programs/compressed-pda/src/instructions.rs:53]:
```suggestion
    // TODO: move to plausibility check
```

File: programs/compressed-pda/src/verify_state.rs [hunk 3]
```diff
@@ -152,59 +152,105 @@ pub fn sum_check(
     }
 }
 
+/// If signer seeds are not provided, invoking program is required.
+/// If invoking program is provided signer seeds are required.
```
Comment (SwenSchaeferjohann) [inline programs/compressed-pda/src/verify_state.rs:156]: basically always `signer_seeds` and `invoking_program` are required ?

Comment (ananas-block) [reply programs/compressed-pda/src/verify_state.rs:156]: when doing a cpi yes

**Review by SwenSchaeferjohann — APPROVED** [review]

**Pull request merged**